File: frontend/src/models/LiquidityMap.ts
export type LiquidityLevelName =
  | "PDH"
  | "PDL"
  | "PWH"
  | "PWL"
  | "H4_SWING_HIGH"
  | "H4_SWING_LOW";

export type LiquiditySide = "BSL" | "SSL";

export interface LiquidityLevel {
  name: LiquidityLevelName;
  side: LiquiditySide;
  level: number;
  distance: number;
  taken: boolean;
}

export interface LiquidityTakenSummary {
  buySideTaken: string[];
  sellSideTaken: string[];
}

export interface LiquidityMap {
  pdh: LiquidityLevel;
  pdl: LiquidityLevel;
  pwh: LiquidityLevel;
  pwl: LiquidityLevel;
  h4SwingHigh: LiquidityLevel;
  h4SwingLow: LiquidityLevel;
  liquidityTaken: LiquidityTakenSummary;
  nearestBuySideLiquidity: LiquidityLevel;
  nearestSellSideLiquidity: LiquidityLevel;
}

File: backend/app/engines/liquidity/LiquidityMap.py
from __future__ import annotations

from typing import Literal
from typing import TypedDict


class LiquidityLevel(TypedDict):
    name: Literal["PDH", "PDL", "PWH", "PWL", "H4_SWING_HIGH", "H4_SWING_LOW"]
    side: Literal["BSL", "SSL"]
    level: float
    distance: float
    taken: bool


class LiquidityTakenSummary(TypedDict):
    buySideTaken: list[str]
    sellSideTaken: list[str]


class LiquidityMap(TypedDict):
    pdh: LiquidityLevel
    pdl: LiquidityLevel
    pwh: LiquidityLevel
    pwl: LiquidityLevel
    h4SwingHigh: LiquidityLevel
    h4SwingLow: LiquidityLevel
    liquidityTaken: LiquidityTakenSummary
    nearestBuySideLiquidity: LiquidityLevel
    nearestSellSideLiquidity: LiquidityLevel
